http_requests.py

```diff
@@ -221,3 +226,66 @@ def delete(self, url: str, return_type=None, **kwargs):
                 return_type=None,
                 **kwargs
         )
+
+
+class Encoder(enum.StrEnum):
+    GZIP = 'gzip'
+
+
+def encode_request(
+    data: str | bytes | dict | typing.IO=None,
+    json: dict=None,
+    headers: dict[str, str]=None,
+    encoder: Encoder=Encoder.GZIP,
+) -> (
+    tuple[collections.abc.Generator[bytes, None, None], dict[str, str]]
+    | collections.abc.Generator[bytes, None, None]
+):
+    '''
+    Encodes the given `data` or `json` property based on the selected `encoder`. If both `data` and
+    `json` are provided, the `data` property is used (same behaviour as implemented by the
+    `requests` library).
+
+    The corresponding `Content-Encoding` header is patched-in to the provided headers dictionary.
+
+    If `headers` is provided and not `None`, the response is a tuple of the compression result and
+    the patched headers, otherwise only the compression result is returned.
+    '''
+    if not data and json is not None:
+        data = js.dumps(json)
+
+    def _encode(obj, encoding: str='utf-8') -> bytes:
+        if isinstance(obj, bytes):
+            return obj
+        elif isinstance(obj, str):
+            return obj.encode(encoding=encoding)
+        elif isinstance(obj, dict):
+            return js.dumps(obj).encode(encoding=encoding)
+        else:
+            raise NotImplementedError(f'Encoding of type {type(obj)} is not (yet) supported')
+
+    def _compress(data, encoder) -> collections.abc.Generator[bytes, None, None]:
+        if isinstance(data, io.BufferedIOBase):
+            if encoder == Encoder.GZIP:
+                compressor = zlib.compressobj(wbits=31)
+                data.seek(0)
+
+                while (chunk := data.read(4096)):
+                    yield compressor.compress(chunk)
+                yield compressor.flush()
+
+        elif hasattr(data, '__iter__') and not isinstance(data, (str, bytes, dict)):
+            raise NotImplementedError(f'Encoding of iterable {type(data)} is not (yet) supported')
+
+        else:
+            data = _encode(data)
+
+            if encoder == Encoder.GZIP:
+                compressor = zlib.compressobj(wbits=31)
+                yield compressor.compress(data) + compressor.flush()
+
+    if headers is not None:
```

Comment:
can't we just check whether headers is truthy?

Reply:
No, because we also want to add the `Content-Encoding` header in case an empty header/dictionary is passed-in.